lib/spack/spack/environment.py

```diff
@@ -1248,6 +1328,52 @@ def _add_concrete_spec(self, spec, concrete, new=True):
         self.concretized_order.append(h)
         self.specs_by_hash[h] = concrete
 
+    def _spec_needs_overwrite(self, spec):
+        # Overwrite the install if it's a dev build (non-transitive)
+        # and the code has been changed since the last install
+        # or one of the dependencies has been reinstalled since
+        # the last install
+
+        # if it's not installed, we don't need to overwrite it
+        if not spec.package.installed:
+            return False
+
+        # if spec and all deps aren't dev builds, we don't need to overwrite it
+        if not any(spec.satisfies(c)
+                   for c in ('dev_path=any', '^dev_path=any')):
+            return False
+
+        # if any dep needs overwrite, or any dep is missing and is a dev build
+        # then overwrite this package
+        if any(
+            self._spec_needs_overwrite(dep) or
+            ((not dep.package.installed) and dep.satisfies('dev_path=any'))
+            for dep in spec.traverse(root=False)
+        ):
+            return True
+
+        # if it's not a direct dev build and its dependencies haven't
+        # changed, it hasn't changed. Don't use satisfaction here because if it
+        # exists, then we need the value later
```

Comment:
It means: you can't just check `spec.satisfies('dev_path=any')`, because that doesn't tell you what the value of `dev_path` is. I'll clarify the comment.